src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1198,14 +1220,68 @@ public bool StatusCanHaveBody(int statusCode)
                    statusCode != 304;
         }
 
-        private void ThrowResponseAlreadyStartedException(string value)
+        private void RejectNonBodyTransferEncodingResponse(bool appCompleted)
         {
-            throw new InvalidOperationException(value + " cannot be set, response had already started.");
+            var ex = BadHttpResponseException.GetException(ResponseRejectionReasons.TransferEncodingSetOnNonBodyResponse, StatusCode.ToString());
+            if (!appCompleted)
+            {
+                // Back out of header creation surface exeception in user code
+                _requestProcessingStatus = RequestProcessingStatus.RequestStarted;
+                throw ex;
+            }
+            else
+            {
+                ReportApplicationError(ex);
+                // 500 Internal Server Error
+                ErrorResetHeadersToDefaults(statusCode: 500);
+            }
+        }
+
+        private void ErrorResetHeadersToDefaults(int statusCode)
```

Comment:
nit: I'd prefer a different name. I'm thinking of `PrepareErrorResponse` or `SetErrorResponse`.